### ISSUE TITLE
Running target operation for training in C++

### ISSUE THREAD

**Issue by Romit-Maulik** [Neargye/hello_tf_c_api#24]
Hello! Before I proceed - I would like to thank you for this excellent repository. I have managed to set up an inference ecosystem in C++ with far less pain than I imagined. I am currently attempting to run a very simple training operation in C++ and was having some trouble. 

I started by looking at the function definition of `TF_SessionRun()` which is
```
void TF_SessionRun(TF_Session* session, const TF_Buffer* run_options,
                   const TF_Output* inputs, TF_Tensor* const* input_values,
                   int ninputs, const TF_Output* outputs,
                   TF_Tensor** output_values, int noutputs,
                   const TF_Operation* const* target_opers, int ntargets,
                   TF_Buffer* run_metadata, TF_Status* status)
```
Accordingly I devised the following code to run my training step
```
// Some tensorflow pointer requirements
TF_Status* status_ = TF_NewStatus();
TF_SessionOptions* options_ = TF_NewSessionOptions();
TF_Session* sess_ = TF_NewSession(graph_, options_, status_);

// This is where the code breaks
const TF_Operation* const* train_ = TF_GraphOperationByName(graph_, "train_step"); 
// Inputs and targets
TF_Tensor* output_tensor_ = tf_utils::CreateTensor(TF_FLOAT,
                                                  output_dims.data(), output_dims.size(),
                                                  target_vals.data(), target_vals.size() * sizeof(float));
TF_Tensor* input_tensor_ = tf_utils::CreateTensor(TF_FLOAT,
                                                  input_dims.data(), input_dims.size(),
                                                  input_vals.data(), input_vals.size() * sizeof(float));
TF_SessionRun(sess_, nullptr, // Run options.
	                &input_op_, &input_tensor_, 1, // Input tensors, input tensor values, number of inputs.
	                &out_op_, &output_tensor_, 1, // Output tensors, output tensor values, number of outputs.
	                train_, 1, // Target operations, number of targets.
	                nullptr, // Run metadata.
	                status_ // Output status.
	                );
printf("Performing backpropagation within C++");
```
Now I know that that `TF_GraphOperationByName` returns a pointer to a `TF_Operation` but the target operation in the `TF_SessionRun` function asks for `const TF_Operation* const*` (a pointer to a constant pointer to a constant `TF_Operation` by my reckoning. Do you have any ideas/examples on how to tackle this issue? 

This would be super helpful! Thanks :)

**Comment by Neargye** [comment]
Try this

```
// Some tensorflow pointer requirements
TF_Status* status_ = TF_NewStatus();
TF_SessionOptions* options_ = TF_NewSessionOptions();
TF_Session* sess_ = TF_NewSession(graph_, options_, status_);

TF_Operation* train_ = TF_GraphOperationByName(graph_, "train_step"); 
std::vector<TF_Operation*> target_opers = {train_ };
// Inputs and targets
TF_Tensor* output_tensor_ = tf_utils::CreateTensor(TF_FLOAT,
                                                  output_dims.data(), output_dims.size(),
                                                  target_vals.data(), target_vals.size() * sizeof(float));
TF_Tensor* input_tensor_ = tf_utils::CreateTensor(TF_FLOAT,
                                                  input_dims.data(), input_dims.size(),
                                                  input_vals.data(), input_vals.size() * sizeof(float));
TF_SessionRun(sess_, nullptr, // Run options.
	                &input_op_, &input_tensor_, 1, // Input tensors, input tensor values, number of inputs.
	                &out_op_, &output_tensor_, 1, // Output tensors, output tensor values, number of outputs.
	                target_opers.data(), 1, // Target operations, number of targets.
	                nullptr, // Run metadata.
	                status_ // Output status.
	                );
printf("Performing backpropagation within C++");
```

**Comment by Neargye** [comment]
TF_SessionRun takes an array of target, so the type const TF_Operation* const*.

**Comment by Romit-Maulik** [comment]
Hi Neargye - thanks! I managed to fix the issue by doing the following:
```
TF_Operation* train_ = TF_GraphOperationByName(graph_, "train_step");
TF_SessionRun(sess_, nullptr, // Run options.
	                &input_op_, &input_tensor_, 1, // Input tensors, input tensor values, number of inputs.
	                &out_op_, &output_tensor_, 1, // Output tensors, output tensor values, number of outputs.
	                &train_, 1, // Target operations, number of targets.
	                nullptr, // Run metadata.
	                status_ // Output status.
	                );
```
which I suspect allowed a safe cast. 

While I have you here - I was trying to use the above code to train an unfrozen graph in C++ (I had defined the graph through the tensorflow python API). However, I noticed that unfrozen graphs cannot be used for any sort of output- (for instance to print the loss). Have you had any experience with this issue and is there a work around for it? 

I suspect I could dump a checkpoint graph to disk and assess on the side - but in-situ would be nice. Do you have an example that shows how to save a graph to disk from C++? Thanks!

**Comment by Neargye** [comment]
Unfortunately, I did not work with ungrozen graph.  I usually use python to train, and use on c ++ then to run.

**Comment by Romit-Maulik** [comment]
I see - thanks! That's what I started out with and am now toying with the idea of moving more functionality into C++. 

PS - I found a stackoverflow [answer](https://stackoverflow.com/questions/37508771/how-to-save-and-restore-a-tensorflow-graph-and-its-state-in-c/37671613#37671613) that details (a bit about how one may save a graph to a checkpoint in C++) - I'll look into it.

**Comment by fullprocess** [comment]
What is target operations? I thought that the session (which contains the graph) included all operations. Why would TF_SessionRun take the session as the first argument AS WELL AS a list of all the target operations?

**Comment by Romit-Maulik** [comment]
The target operations here correspond to those you want to run during the session on the _unfrozen_ graph. In the previous snippet they were those related to optimization of the graph. This gist has all the code you'll need to understand this!

https://gist.github.com/asimshankar/7c9f8a9b04323e93bb217109da8c7ad2

**Comment by fullprocess** [comment]
Thanks Romit! So if I understand correctly, if you want to include all graph operations you leave Target ops as NULL? I am using a frozen graph of Deeplab pre-trained on the Cityscapes dataset, so I believe it is fully optimized and ready to predict.

**Comment by Romit-Maulik** [comment]
Yup! So for a pretrained and frozen graph you would use something like 
```
// Inputs and dimensions
float input_vals[num_samples][num_inputs]; // populate this with inputs
const std::vector<std::int64_t> input_dims = {num_samples, num_inputs}; //dimensions
// Set up tensors
TF_Tensor* output_tensor_ = nullptr;
TF_Tensor* input_tensor_ = tf_utils::CreateTensor(TF_FLOAT,
                                          input_dims, input_dims.size(),
                                          &input_vals, num_samples*num_inputs*sizeof(float));
// Arrays of tensors
TF_Tensor* input_tensors_[1] = {input_tensor_}; // Array of all the inputs to the model
TF_Tensor* output_tensors_[1] = {output_tensor_}; // Array of all the outputs from the model
// Arrays of operations
TF_Output inputs[1] = {input_ph_}; // The input placeholder(s) 
TF_Output outputs[1] = {output_}; // The output operation

TF_SessionRun(sess_,
                nullptr, // Run options.
                inputs, input_tensors_, 1, // Input tensor ops, input tensor values, number of inputs.
                outputs, output_tensors_, 1, // Output tensor ops, output tensor values, number of outputs.
                nullptr, 0, // *No* target operations, number of target ops.
                nullptr, // Run metadata.
                status_ // Output status.
);
```

**Comment by yangjituan** [comment]
Hi ,
i want to do gradient descent  in tensorflow-C_api ,but can't find any example tell me how to code,
such as :
      how to set and initialize a variable  ?
      how to set optimizer and train?
@Romit-Maulik @Neargye Thanks for your attention and  hope to receive your reply.

**Comment by Neargye** [comment]
@yangjituan you can also check https://gist.github.com/asimshankar/7c9f8a9b04323e93bb217109da8c7ad2.

**Comment by yangjituan** [comment]
@Neargye Thanks very much, It's  very useful,
